### PR TITLE
Add features to SystemOutage and SystemOutageTask

### PR DESCRIPTION
**Pull request by jinnnn9**

### CODE REVIEW

File: src/foam/nanos/so/rules.jrl [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  "action":{
+    "class":"foam.nanos.so.task.TaskInitRuleAction"
+  },
+  "enabled":true,
+  "lifecycleState":1
+})
+
+p({
+  "class":"foam.nanos.ruler.Rule",
+  "id":"341E0A0C-3BA3-4C07-83C1-B9313DCFE001",
```
Comment (jlhughes) [inline src/foam/nanos/so/rules.jrl]: please use package naming for rule ids
foam_nanos_so_task_TaskCleanupOnRemovalRule
for example.

File: src/foam/nanos/so/task/TaskInitRuleAction.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    'java.util.UUID'
+  ],
+
+  methods: [
+    {
+      name: 'applyAction',
+      javaCode: `
+        SystemOutage so = (SystemOutage) obj;
+        for ( SystemOutageTask task : so.getTasks() ) {
+          if ( task.getId().isEmpty() ) {
```
Comment (jlhughes) [inline src/foam/nanos/so/task/TaskInitRuleAction.js]: This first part could be replaced with a factory on SystemOutageTask id. 
      javaFactory: `
        return UUID.randomUUID().toString();
      `,
but I suppose if you already have all this rule logic, it doesn't matter

File: src/foam/nanos/so/SystemOutage.js [hunk 3]
```diff
@@ -109,6 +112,23 @@ foam.CLASS({
...
         }
       `
+    },
+    {
+      javaType: 'SystemOutageTask[]',
+      name: 'findNonOverlappingTasks',
+      documentation: 'Returns a list of tasks whose tasks are in \'this\' but not in other',
+      args: 'X x, SystemOutage other',
+      javaCode: `
+        List<String> otherTaskIds = Arrays.asList(other.getTasks())
```
Comment (jlhughes) [inline src/foam/nanos/so/SystemOutage.js]: I'm pretty sure Kevin has one of these in foam already.

File: src/foam/nanos/so/pom.js [hunk 4]
```diff
@@ -13,6 +13,14 @@ foam.POM({
       flags: "js|java" },
     { name: "SystemOutageTask",
       flags: "js|java" },
+    { name: "task/TaskInitPredicate",
```
Comment (jlhughes) [inline src/foam/nanos/so/pom.js]: alphabetize please, so move all the task/ after the last System.
and if we are going with this extra directory, then SystemOutageTask should move there as well. 
and the other existing tasks. 
but, all development will be of tasks, so the top level so directory will have a half a dozen files and everything else will be in task/  so perhaps task isn't needed. 
give it some thought.

Comment (jinnnn9) [reply src/foam/nanos/so/pom.js]: will remove task directory

File: src/foam/nanos/so/SystemOutageTask.js [hunk 5]
```diff
@@ -31,6 +54,36 @@ foam.INTERFACE({
     {
       name: 'deactivate',
       args: 'X x'
+    },
+    {
+      name: 'cleanUp',
+      args: 'X x',
+      documentation: 'work to be done when task is removed',
+      javaCode: `
+        deactivate(x);
```
Comment (jlhughes) [inline src/foam/nanos/so/SystemOutageTask.js]: Should we test SystemOutage.active first
so if active, call deactivate, else it should have already been called.

File: src/foam/nanos/so/SystemOutage.js (hunk 3 above)
Comment (jlhughes) [reply src/foam/nanos/so/SystemOutage.js]: We don't.

File: src/foam/nanos/so/SystemOutageTask.js [hunk 6]
```diff
@@ -19,8 +19,31 @@ foam.INTERFACE({
...
+  properties: [
+    {
+      class: 'String',
+      name: 'id',
+      order: -20,
+      visibility: 'RO'
+    },
+    {
+      class: 'Reference',
+      of: 'foam.nanos.so.SystemOutage',
```
Comment (jlhughes) [inline src/foam/nanos/so/SystemOutageTask.js]: Thinking, how about another interface  SystemOutageRuleTask which is rule based. 
and add cleanup, save and such to it. 
then we have a simple task for messages and such, and another that is rule based.

File: src/foam/nanos/so/pom.js (hunk 4 above)
Comment (jinnnn9) [reply src/foam/nanos/so/pom.js]: done

File: src/foam/nanos/so/SystemOutageTask.js (hunk 6 above)
Comment (jinnnn9) [reply src/foam/nanos/so/SystemOutageTask.js]: discussed in a chat and decided to keep it as it is as other tasks might need these features

File: src/foam/nanos/so/rules.jrl (hunk 1 above)
Comment (jinnnn9) [reply src/foam/nanos/so/rules.jrl]: done

File: src/foam/nanos/so/task/TaskInitRuleAction.js (hunk 2 above)
Comment (jinnnn9) [reply src/foam/nanos/so/task/TaskInitRuleAction.js]: will add a relationship between system outage and task so this init rule will no longer be needed

**Review by jlhughes — APPROVED** [review]

**Review by jlhughes — APPROVED** [review]

**Pull request merged**